features/netsocket/PPPInterface.h

```diff
@@ -0,0 +1,197 @@
+     */
+    virtual const char *get_netmask();
+
+    /** Get the local gateways
+     *
+     *  @return         Null-terminated representation of the local gateway
+     *                  or null if no network mask has been received
+     */
+    virtual const char *get_gateway();
+
+    /** Get the network interface name
+     *
+     *  @return         Null-terminated representation of the network interface name
+     *                  or null if  interface not exists
+     */
+    virtual char *get_interface_name(char *interface_name);
+
+    /** Set the network interface as default one
+      */
+    virtual void set_as_default();
+
+    /** Register callback for status reporting
+     *
+     *  @param status_cb The callback for status changes
+     */
+    virtual void attach(mbed::Callback<void(nsapi_event_t, intptr_t)> status_cb);
+
+    /** Get the connection status
+     *
+     *  @return         The connection status according to nsapi_connection_status_t
+     */
+    virtual nsapi_connection_status_t get_connection_status() const;
+
+    /** Set blocking status of connect() which by default should be blocking
+     *
+     *  @param blocking true if connect is blocking
+     *  @return         0 on success, negative error code on failure
+     */
+    virtual nsapi_error_t set_blocking(bool blocking);
+
+    /** Sets file stream used to communicate with modem
+     *
+     * @param stream Pointer to file handle
+     */
+    virtual void set_stream(mbed::FileHandle *stream);
+
+    /** Sets IP protocol versions of IP stack
+     *
+     * @param ip_stack IP protocol version
+     */
+    virtual void set_ip_stack(nsapi_ip_stack_t ip_stack);
+
+    /** Sets user name and password for PPP protocol
+     *
+     * @param uname    User name
+     * @param password Password
+     */
+    virtual void set_credentials(const char *uname, const char *password);
+
+    /** Provide access to the PPP
+     *
+     * This should be used with care - normally the network stack would
+     * control the PPP, so manipulating the PPP while the stack
+     * is also using it (ie after connect) will likely cause problems.
+     *
+     * @return          Reference to the PPP in use
+     */
+    PPP &getppp() const
+    {
+        return _ppp;
+    }
+
+    virtual PPPInterface *pppInterface()
+    {
+        return this;
+    }
+
+protected:
+    /** Provide access to the underlying stack
+     *
+     *  @return The underlying network stack
+     */
+    virtual NetworkStack *get_stack();
+
+    PPP &_ppp;
+    OnboardNetworkStack &_stack;
+    OnboardNetworkStack::Interface *_interface;
+    bool _dhcp;
+    bool _blocking;
+    char _ip_address[NSAPI_IPv6_SIZE];
+    char _netmask[NSAPI_IPv4_SIZE];
+    char _gateway[NSAPI_IPv4_SIZE];
+    mbed::Callback<void(nsapi_event_t, intptr_t)> _connection_status_cb;
+
+    mbed::FileHandle *_stream;
+    nsapi_ip_stack_t _ip_stack;
+    char _uname[30];
```

Comment:
Should a define be used instead?

Reply:
Chained user name and password to pointers inside ppp interface class (similar to cellular interface).